### PR TITLE
Fix Tpetra with MemorySpace::Default

### PR DESCRIPTION
**Pull request by masterleinad**
Trying compiling the TpetraWrappers with `MemorySpace::Default` (and a GPU backend) required adding the `MemorySpace` template argument in a bunch of places.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
Is this a candidate for the 9.6 release branch?

**Comment by masterleinad** [comment]
> Is this a candidate for the 9.6 release branch?

That might be a good idea.

File: include/deal.II/base/index_set.h [hunk 1]
```diff
@@ -608,11 +612,17 @@ class IndexSet
                     const bool     overlapping  = false) const;
 
 #  ifdef DEAL_II_TRILINOS_WITH_TPETRA
-  Tpetra::Map<int, types::signed_global_dof_index>
+  template <
+    typename NodeType =
+      LinearAlgebra::TpetraWrappers::TpetraTypes::NodeType<MemorySpace::Host>>
```
Comment (tjhei) [inline include/deal.II/base/index_set.h:617]: why do you pick Host as the default here?

File: include/deal.II/lac/read_write_vector.h [hunk 2]
```diff
@@ -781,7 +782,8 @@ namespace LinearAlgebra
      * Return a TpetraWrappers::CommunicationPattern and store it for future
      * use.
      */
-    TpetraWrappers::CommunicationPattern
+    template <typename MemorySpace = dealii::MemorySpace::Host>
```
Comment (tjhei) [inline include/deal.II/lac/read_write_vector.h:785]: same here

**Review by tjhei — APPROVED** [review]
While nice to have in 9.6, it has some chance of producing compile/errors warnings in some configurations.

File: include/deal.II/base/index_set.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/index_set.h:617]: We are using `MemorySpace::Host` as the default for all `Tpetra` wrappers. Without the default argument, we would need to adapt a couple of tests.

**Comment by masterleinad** [comment]
> While nice to have in 9.6, it has some chance of producing compile/errors warnings in some configurations.

I would rather have the interfaces in `IndexSet` fixed in this release so that we don't have to break it again (although it should work with the default template arguments).

**Comment by tjhei** [comment]
> I would rather have the interfaces in `IndexSet` fixed in this release

makes sense. :+1:

**Review by tjhei — APPROVED** [review]

**Review by tjhei — APPROVED** [review]

**Pull request merged**